### ISSUE TITLE
[Question] Missing state / province details for few of countries

### ISSUE THREAD

**Issue by mahesh-chotrani** [amcharts/amcharts4#3200]
**Question**

Following this example, could have state / province drill down maps implemented
https://www.amcharts.com/docs/v4/tutorials/building-drill-down-maps/

Found that there are few countries, for which state / province details are missing, few of these being Uganda, Jamaica, Taiwan, Cuba, Jordan, Uruguay (_and there are more_). 

I could not find JSON for few of these as well at the location https://cdn.amcharts.com/lib/4/geodata/json 

Would like to confirm if there is anything I have missed out while looking for details of these countries. 

Thanks

**Comment by martynasma** [comment]
> Found that there are few countries, for which state / province details are missing, few of these being Uganda, Jamaica, Taiwan, Cuba, Jordan, Uruguay (and there are more).

Correct. We do not have maps for all of the countries. We're adding them constantly. Make sure you follow [this repo](https://github.com/amcharts/amcharts4-geodata) for the updates.

BTW, we do have maps of Jamaica, Uruguay, and Cuba.

**Comment by mahesh-chotrani** [comment]
Thanks for confirming @martynasma

**Comment by mahesh-chotrani** [comment]
While we are implementing drill down feature for one of our customers (following example https://www.amcharts.com/demos/drill-down-to-countries/), found some information, which may be helpful to have updated without much effort, so thought to share with you. 

For the following countries that are mentioned in https://www.amcharts.com/lib/4/geodata/json/data/countries2.json , have empty maps for province file names, but corresponding province files exist:

 ```
 Country name Afghanistan, filename afghanistanLow
 Country name British Virgin Islands, filename britishVirginIslandsLow
 Country name Cuba, filename cubaLow
 Country name Falkland Islands, filename falklandIslandsLow
 Country name Grenada, filename grenadaLow
 Country name Guadeloupe, filename guadeloupeLow
 Country name Guam, filename guamLow
 Country name Guatemala, filename guatemalaLow
 Country name Haiti, filename haitiLow
 Country name Jamaica, filename jamaicaLow
 Country name Jordan, filename jordanLow
 Country name Lebanon, filename lebanonLow
 Country name Martinique, filename martiniqueLow
 Country name Montserrat, filename montserratLow
 Country name Paraguay, filename paraguayLow
 Country name Trinidad and Tobago, filename trinidadTobagoLow
 Country name Uruguay, filename uruguayLow
 Country name US Virgin Islands, filename usVirginIslandsLow
```

And in https://www.amcharts.com/lib/4/geodata/json/data/countries.json  (_a short version of countries2.json_) those countries are not mentioned at all.

Please ignore if you are already aware of this and done for a purpose.

**Comment by martynasma** [comment]
Thanks. We'll need to update those data files.

**Comment by mahesh-chotrani** [comment]
@martynasma  Another thing noticed, so thought to update (please ignore, if you are already aware of this) - few of the countries have state / province code which is numeric, while majority have a 2 or 3 letter alphabet. 
Though, state / province maps are rendered without any problems. 

For example, Japan, Italy, South Korea have state codes which are all numeric. Where as majority countries (e.g., US, Canada, India, etc.) have state codes as 2 letter alphabets. (_in both cases, prefixed by 2 letter country code and a hyphen_)

Not sure, if this is as per what a country maintains by itself or a serial number.

**Comment by martynasma** [comment]
We are following ISO identifications where those are available. In other cases we follow how countries themselves identify their sub-divisions.

For example in the US states have their letter-based codes, but congressional districts have numeric IDs.

In any case, if you find some maps use outdated or incorrect IDs, you can let us know by opening separate issue.

**Comment by sashaepstein** [comment]
Hi @martynasma, ISO defines both, alphabetical and numerical codes, for example for Italy:

```
Code | Subdivision name | In region
IT-AL | Alessandria     | 21
IT-AN | Ancona          | 57
```
We've been using your json files not only to get the id element (like "id":"IT-21") for amcharts province map, but also as a source for the province alphabetical codes (that in this example would be the right part of the id - AL).
Is it possible that you add a new element in json with the province alphabetical code?

**Comment by martynasma** [comment]
Well, out Italian map shows regions, not provinces.

I don't think regions have alpha notations, like provinces do.

We'll consider creating a province map version of Italy.

**Comment by sashaepstein** [comment]
In some countries there are multiple sets of divisions (like regions and provinces in Italy, and provinces and districts in Serbia, etc.). Some of those divisions have only numerical codes and other have alphabetical or alphanumerical codes that can be used by that country in the mailing addresses. The problem we are facing is to correlate whatever division postal code a particular country is using in their mailing address with some data provided in your json files to get the ID element for amcharts province map. I would expect other customers have a similar problem.

**Comment by martynasma** [comment]
Well, we have used whatever ISO codes were in effect at the moment of creation of the map. Some of those maps are a few years old, and might be outdated.

If you find any such outdated information, please do let us know, and we'll do our best to update.

In case of Italian maps, our map is for regions, which use numeric IDs.

We're currently looking into producing a separate province map, which would use letter-based IDs.

**Comment by sashaepstein** [comment]
>If you find any such outdated information, please do let us know

Certainly. Here is one:  for Great Britain the ISO_3166-2:GB document defines the following divisions: 4 "provinces", 3 "nations" and several dozen of "second-level subdivisions". The json file ukLow/ukHigh contains only 3 of the "provinces", none of the "nations" and "second-level subdivisions", but contains a dozen or so of codes not mentioned in ISO. I found most of "second-level subdivisions" in json file ukCounties but the 2 char codes there don't match 3 char codes in ISO.
Also, it would be nice to have all codes mentioned in ISO in a single json file.
 
>We're currently looking into producing a separate province map, which would use letter-based IDs.

Can those letter-based IDs be used to produce amcharts "province map"?

**Comment by martynasma** [comment]
We have 3 different versions of the UK map:

https://github.com/amcharts/amcharts4/blob/master/dist/geodata/script/json/ukCountiesLow.json
https://github.com/amcharts/amcharts4/blob/master/dist/geodata/script/json/ukCountriesLow.json
https://github.com/amcharts/amcharts4/blob/master/dist/geodata/script/json/ukLow.json

At this moment we do not plan on adding any other versions.

We'll review if the ISO coding is up-to-date on those.

**Comment by martynasma** [comment]
OK, so I have reviewed what we have.

`ukCountries` map, uses 3-letter ISO identifiers:
https://en.wikipedia.org/wiki/ISO_3166-2:GB#Countries_and_province

`ukCounties` uses 2-letter identifiers, instead of 3-letter ISO ones, we'll need to investigate this.

`uk` uses UK's NUTS 3-letter codes for regions like North East, etc:
http://www.statoids.com/ugb.html

**Comment by sashaepstein** [comment]
>ukCountries map, uses 3-letter ISO identifiers

It has the following problems: it misses GB-ENG defined in ISO, and it has plenty of 2-letter codes that are **not** defined in ISO.

>We'll consider creating a province map version of Italy.

Please do. I spoke to Italian people and they confirmed that they use their provinces in the mail addresses.

**Comment by martynasma** [comment]
> > ukCountries map, uses 3-letter ISO identifiers

> It has the following problems: it misses GB-ENG defined in ISO, and it has plenty of 2-letter codes that are not defined in ISO.

Are you sure? It does have GB-ENG.

**Comment by sashaepstein** [comment]
My apologies, I confused the file name: _countries_ and _counties_ are too close :-)
You are correct,  _countries_ one does contain all 4 codes for provinces from ISO. However only one of them has "TYPE":"Province", the other 3 have "TYPE":"Country", along with entries for real countries like Ireland:
```
"id":"IE",
"name":"Ireland",
"TYPE":"Country",
"CNTRY":"Ireland"
```
This makes it difficult to use _ukCountries.json_ map because it mixes UK provinces and countries in the same file and TYPE doesn't help to identify which ID should be taken for "province code" and which ID should be ignored because it is for real country, not province. 
That is why I use  _ukCounties.json_ map that doesn't mix countries and provinces, but as I mentioned above, it contains only 3 out of 4 provinces defined in ISO and has 2-letter ids for other divisions.

**Comment by martynasma** [comment]
If `ukCountries` map is what you need just with extra countries, you can easily hide those out of use using `exclude`:

https://www.amcharts.com/docs/v4/chart-types/map/#Including_and_excluding_objects

**Comment by martynasma** [comment]
OK, so we have revisited this. What we have for UK counties map is a map of historic counties, using HASC codes, that seem very much still in popular use.

The new counties (with ISO IDs) are not as popular. We might at some point make the new ISO map, though. Just don't have the ETA.

**Comment by sashaepstein** [comment]
>In case of Italian maps, our map is for regions, which use numeric IDs.

Italy province map is incomplete, it is missing all provinces that are in use there.
Also why Corse is in Italy province map if it belongs to France:
`"name":"Corsica","id":"FR-H","NAME_ENG":"Corsica","CNTRY":"France"`

**Comment by martynasma** [comment]
We're working on a new map for Italian provinces.

Corsica is there for reference, because otherwise the island group would look weird.

**Comment by martynasma** [comment]
Just released geodata 4.1.20. It contains a province map of Italy.

**Comment by sashaepstein** [comment]
I see 2 new json files `italyProvincesLow.json` and `italyProvincesHighy`.json for Italy provinces but `data/countries.json` and  `data/countries2.json` don't mention those files. Without this reference the newly added file can't be pulled. Please update.
